File: src/data/ua/ksn/fmg/controller/serializable/PlayersModel.java
package ua.ksn.fmg.controller.serializable;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.Externalizable;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.ObjectInput;
import java.io.ObjectInputStream;
import java.io.ObjectOutput;
import java.io.ObjectOutputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

import ua.ksn.crypt.Simple3DES;
import ua.ksn.fmg.controller.event.PlayerModelEvent;
import ua.ksn.fmg.controller.event.PlayerModelListener;
import ua.ksn.fmg.controller.types.ESkillLevel;
import ua.ksn.fmg.controller.types.User;
import ua.ksn.fmg.model.mosaics.EMosaic;

/** хранилище пользователей и их игровой статистики */
public class PlayersModel implements Externalizable {
	//private static final long version = Main.serialVersionUID;
	private final long version;
	
	public PlayersModel(long version) { this.version = version; }

	private class Record implements Externalizable {
		private User user;
		private StatisticCounts statistics[][] = new StatisticCounts[EMosaic.values().length][ESkillLevel.values().length-1];

		/** new User */
		public Record(User user) {
			this.user = user;
			for (EMosaic mosaic : EMosaic.values())
				for (ESkillLevel skill : ESkillLevel.values())
					if (skill == ESkillLevel.eCustom)
						continue;
					else
						statistics[mosaic.ordinal()][skill.ordinal()] = new StatisticCounts();
		}
		/** from file */
		public Record(ObjectInput in) throws IOException {
			try {
				for (EMosaic mosaic : EMosaic.values())
					for (ESkillLevel skill : ESkillLevel.values())
						if (skill == ESkillLevel.eCustom)
							continue;
						else
							statistics[mosaic.ordinal()][skill.ordinal()] = new StatisticCounts();
				readExternal(in);
			} catch (ClassNotFoundException ex) {
				throw new RuntimeException(ex);
			}
		}

		@Override
		public void readExternal(ObjectInput in) throws IOException, ClassNotFoundException {
			user = new User(in);
			for (StatisticCounts[] record: statistics)
				for (StatisticCounts subRecord: record)
					subRecord.readExternal(in);
		}
		@Override
		public void writeExternal(ObjectOutput out) throws IOException {
			user.writeExternal(out);
			for (StatisticCounts[] record: statistics)
				for (StatisticCounts subRecord: record)
					subRecord.writeExternal(out);
		}
		@Override
		public String toString() {
			return user.getName();
		}
	}

	private List<PlayersModel.Record> players = new ArrayList<PlayersModel.Record>();

	public void removePlayer(UUID userId) {
		Record rec = find(userId);
		if (rec == null)
			throw new IllegalArgumentException("User " + userId + " not exist");
		int pos = players.indexOf(rec);
		players.remove(pos);

		fireChanged(new PlayerModelEvent(this, pos, PlayerModelEvent.DELETE));
	}
	public boolean isExist(UUID userId) { return find(userId) != null; }
	public int size() { return players.size(); };
	public UUID addNewPlayer(String name, String pass) {
		if ((name == null) || name.isEmpty())
			throw new IllegalArgumentException("Invalid player name. Need not empty.");
		for (Record rec: players)
			if (rec.user.getName().equalsIgnoreCase(name))
				throw new IllegalArgumentException("Please enter a unique name");

		User user = new User(name, pass, null);
		players.add(new PlayersModel.Record(user));
		fireChanged(new PlayerModelEvent(this, players.size()-1, PlayerModelEvent.INSERT));
		return user.getGuid();
	}
	public int indexOf(User user) {
		Record recFind = null;
		for (Record rec: players)
			if (rec.user.getGuid().equals(user.getGuid())) {
				recFind = rec;
				break;
			}
		if (recFind == null)
			return -1;
		return players.indexOf(recFind);
	}

	/**
	 * Установить статистику для игрока
	 * @param userId - идентификатор игрока
	 * @param mosaic - на какой мозаике
	 * @param skill - на каком уровне сложности
	 * @param victory - выиграл ли?
	 * @param countOpenField - кол-во открытых ячеек
	 * @param playTime - время игры
	 * @param clickCount - кол-во кликов
	 */
	public void setStatistic(UUID userId, EMosaic mosaic, ESkillLevel skill, boolean victory, long countOpenField, long playTime, long clickCount) {
		if (skill == ESkillLevel.eCustom)
			return;
		Record rec = find(userId);
		if (rec == null)
			throw new IllegalArgumentException("User " + userId + " not exist");

		StatisticCounts subRec = rec.statistics[mosaic.ordinal()][skill.ordinal()];
		subRec.gameNumber++;
		subRec.gameWin    += victory ? 1:0;
		subRec.openField  += countOpenField;
		if (victory) {
			subRec.playTime   += playTime;
			subRec.clickCount += clickCount;
		}

		int pos = players.indexOf(rec);
		fireChanged(new PlayerModelEvent(this, pos, PlayerModelEvent.CHANGE_STATISTICS, mosaic, skill));
	}
	private Record find(UUID userId) {
		if (userId != null)
			for (Record rec: players)
				if (rec.user.getGuid().equals(userId))
					return rec;
		return null;
	}
	public User getUser(int pos) {
		if ((pos < 0) || (pos>=players.size()))
			throw new IllegalArgumentException("Invalid position " + pos);
			//return null;
		return players.get(pos).user;
	}
	public User getUser(UUID userId) {
		Record rec = find(userId);
		if (rec == null)
			throw new IllegalArgumentException("User " + userId + " not exist");
		return rec.user;
	}
	public StatisticCounts getInfo(UUID userId, EMosaic mosaic, ESkillLevel skillLevel) {
		Record rec = find(userId);
		if (rec == null)
			throw new IllegalArgumentException("User " + userId + " not exist");

		try {
			return rec.statistics[mosaic.ordinal()][skillLevel.ordinal()].clone();
		} catch (CloneNotSupportedException ex) {
			throw new RuntimeException(ex);
		}
	}
	public int getPos(UUID userId) {
		if (userId == null)
			return -1;
		Record rec = find(userId);
		if (rec == null)
			return -1;
		return players.indexOf(rec);
	}

	@Override
	public void writeExternal(ObjectOutput out) throws IOException {
		out.writeLong(version);
		out.writeInt(players.size());
		for (PlayersModel.Record rec : players)
			rec.writeExternal(out);
	}

	@Override
	public void readExternal(ObjectInput in) throws IOException, ClassNotFoundException {
		setDefaults();
		if (version != in.readLong())
			throw new RuntimeException("Unknown version");
		int size = in.readInt();
		for (int i=0; i<size; i++)
			players.add(new PlayersModel.Record(in));

		fireChanged(new PlayerModelEvent(this, players.size()-1, PlayerModelEvent.INSERT_ALL));
	}

	private void setDefaults() {
		int len = players.size();
		players.clear();
		fireChanged(new PlayerModelEvent(this, len-1, PlayerModelEvent.DELETE_ALL));
	}

	/**
	 * Load STC data from file
	 * @return <b>true</b> - successful read; <b>false</b> - not exist or fail read, and set to defaults
	 */
	public boolean Load() {
		File file = getStcFile();
		if (!file.exists()) {
			setDefaults();
			return false;
		}

		try {
			// 1. read from file
			ObjectInputStream in = new ObjectInputStream(new FileInputStream(file));
			if (in.readLong() != version)
				throw new RuntimeException("Invalid file data - unknown version");
			byte[] data = new byte[in.readInt()];
			int read = 0;
			do {
				int curr = in.read(data, read, data.length-read);
				if (curr == -1)
					break;
				read += curr;
			} while(read < data.length);
			if (read != data.length)
				throw new IOException("Invalid data length. Ожидалось " + data.length + " байт, а прочитано " + read + " байт.");
			in.close();

			// 2. decrypt data
			try {
				data = new Simple3DES(Long.toString(version)).decrypt(data);
			} catch (Exception ex) {
				throw new RuntimeException(ex);
			}

			// 3. deserializable object
			in = new ObjectInputStream(new ByteArrayInputStream(data));
			this.readExternal(in);

			in.close();

			return true;
		} catch (Exception ex) {
			ex.printStackTrace();
			setDefaults();
			return false;
		}
	}

	public void Save() throws FileNotFoundException, IOException {
		// 1. serializable object
		ByteArrayOutputStream byteRaw = new ByteArrayOutputStream();
		ObjectOutputStream out = new ObjectOutputStream(byteRaw);
		this.writeExternal(out);
		out.flush();

		// 2. crypt data
		byte[] cryptData;
		try {
			cryptData = new Simple3DES(Long.toString(version)).encrypt(byteRaw.toByteArray());
		} catch (Exception ex) {
			throw new RuntimeException(ex);
		}

		// 3. write to file
		out = new ObjectOutputStream(new FileOutputStream(getStcFile()));
		out.writeLong(version); // save version and decrypt key
		int len = cryptData.length;
		out.writeInt(len);
		out.write(cryptData);

		out.flush();
		out.close();
	}

	/**
	 * STatistiCs file name
	 */
	public static File getStcFile() {
		return new File(System.getProperty("user.dir") + System.getProperty("file.separator") + "Mines.stc");
	}

	public void addPlayerListener(PlayerModelListener l) {
		arrPlayerListener.add(l);
	}
	public void removeTableModelListener(PlayerModelListener l) {
		arrPlayerListener.remove(l);
	}
	private List<PlayerModelListener> arrPlayerListener = new ArrayList<PlayerModelListener>();
	private void fireChanged(PlayerModelEvent e) {
		for (PlayerModelListener listener: arrPlayerListener)
			listener.playerChanged(e);
	}
	public void setUserName(int pos, String name) {
		User user = getUser(pos);
		user.setName(name);
		fireChanged(new PlayerModelEvent(this, pos, PlayerModelEvent.UPDATE));
	}
}

File: src/src_data/fmg/data/controller/serializable/PlayersModel.cs
using System;
using System.IO;
using System.Linq;
using System.Text;
using System.Collections.Generic;
using fmg.common;
using fmg.data.controller.Event;
using fmg.data.controller.types;
using fmg.core.types;


namespace fmg.data.controller.serializable {

   /// <summary>хранилище пользователей и их игровой статистики</summary>
   public abstract class PlayersModel : IExternalizable {
      //private static final long version = Main.serialVersionUID;
      private readonly long version;
   
      public PlayersModel(long version) { this.version = version; }

      private class Record : IExternalizable {
         public User user;
         public StatisticCounts[,] statistics = new StatisticCounts[EMosaicEx.GetValues().Length, ESkillLevelEx.GetValues().Length - 1];

         /// <summary>new User</summary>
          public Record(User user) {
            this.user = user;
            foreach (var mosaic in EMosaicEx.GetValues())
               foreach (var skill in ESkillLevelEx.GetValues())
                  if (skill == ESkillLevel.eCustom)
                     continue;
                  else
                     statistics[mosaic.Ordinal(), skill.Ordinal()] = new StatisticCounts();
         }
         /// <summary>from file</summary>
         public Record(BinaryReader input) {
            foreach (var mosaic in EMosaicEx.GetValues())
               foreach (var skill in ESkillLevelEx.GetValues())
                  if (skill == ESkillLevel.eCustom)
                     continue;
                  else
                     statistics[mosaic.Ordinal(), skill.Ordinal()] = new StatisticCounts();
            readExternal(input);
         }

         public void readExternal(BinaryReader input) {
            user = new User(input);
            foreach (StatisticCounts record in statistics)
               record.readExternal(input);
         }
         public void writeExternal(BinaryWriter output) {
            user.writeExternal(output);
            foreach (StatisticCounts record in statistics)
               record.writeExternal(output);
         }

         public override String ToString() {
            return user.Name;
         }
      }

      private IList<PlayersModel.Record> players = new List<PlayersModel.Record>();

      public void removePlayer(Guid userId) {
         Record rec = find(userId);
         if (rec == null)
            throw new ArgumentException("User " + userId + " not exist");
         int pos = players.IndexOf(rec);
         players.Remove(rec);

         fireChanged(new PlayerModelEventArgs(pos, PlayerModelEventArgs.DELETE));
      }
      public bool isExist(Guid userId) { return find(userId) != null; }
      public int Size { get { return players.Count; } }
      public Guid addNewPlayer(String name, String pass) {
         if (string.IsNullOrEmpty(name))
            throw new ArgumentException("Invalid player name. Need not empty.");
         foreach (Record rec in players)
            if (rec.user.Name.Equals(name, StringComparison.OrdinalIgnoreCase))
               throw new ArgumentException("Please enter a unique name");

         User user = new User(name, pass, null);
         players.Add(new PlayersModel.Record(user));
         fireChanged(new PlayerModelEventArgs(players.Count-1, PlayerModelEventArgs.INSERT));
         return user.Guid;
      }
      public int indexOf(User user) {
         Record recFind = null;
         foreach (Record rec in players)
            if (rec.user.Guid.Equals(user.Guid)) {
               recFind = rec;
               break;
            }
         if (recFind == null)
            return -1;
         return players.IndexOf(recFind);
      }

      /// <summary>Установить статистику для игрока</summary>
      /// <param name="userId">идентификатор игрока</param>
      /// <param name="mosaic">на какой мозаике</param>
      /// <param name="skill">на каком уровне сложности</param>
      /// <param name="victory">выиграл ли?</param>
      /// <param name="countOpenField">кол-во открытых ячеек</param>
      /// <param name="playTime">время игры</param>
      /// <param name="clickCount">кол-во кликов</param>
      public void setStatistic(Guid userId, EMosaic mosaic, ESkillLevel skill, bool victory, long countOpenField, long playTime, long clickCount) {
         if (skill == ESkillLevel.eCustom)
            return;
         Record rec = find(userId);
         if (rec == null)
            throw new ArgumentException("User " + userId + " not exist");

         StatisticCounts subRec = rec.statistics[mosaic.Ordinal(), skill.Ordinal()];
         subRec.gameNumber++;
         subRec.gameWin    += victory ? 1:0;
         subRec.openField  += countOpenField;
         if (victory) {
            subRec.playTime   += playTime;
            subRec.clickCount += clickCount;
         }

         int pos = players.IndexOf(rec);
         fireChanged(new PlayerModelEventArgs(pos, PlayerModelEventArgs.CHANGE_STATISTICS, mosaic, skill));
      }
      private Record find(Guid userId) {
         if (userId != null)
            foreach (Record rec in players)
               if (rec.user.Guid.Equals(userId))
                  return rec;
         return null;
      }
      public User getUser(int pos) {
         if ((pos < 0) || (pos>=players.Count))
            throw new ArgumentException("Invalid position " + pos);
            //return null;
         return players[pos].user;
      }
      public User getUser(Guid userId) {
         Record rec = find(userId);
         if (rec == null)
            throw new ArgumentException("User " + userId + " not exist");
         return rec.user;
      }
      public StatisticCounts getInfo(Guid userId, EMosaic mosaic, ESkillLevel skillLevel) {
         Record rec = find(userId);
         if (rec == null)
            throw new ArgumentException("User " + userId + " not exist");

         return rec.statistics[mosaic.Ordinal(), skillLevel.Ordinal()].clone();
      }
      public int getPos(Guid userId) {
         if (userId == null)
            return -1;
         Record rec = find(userId);
         if (rec == null)
            return -1;
         return players.IndexOf(rec);
      }

      public void writeExternal(BinaryWriter output) {
         output.Write(version);
         output.Write(players.Count);
         foreach (PlayersModel.Record rec in players)
            rec.writeExternal(output);
      }

      public void readExternal(BinaryReader input) {
         setDefaults();
         if (version != input.ReadInt64())
            throw new Exception("Unknown version");
         int size = input.ReadInt32();
         for (int i=0; i<size; i++)
            players.Add(new PlayersModel.Record(input));

         fireChanged(new PlayerModelEventArgs(players.Count-1, PlayerModelEventArgs.INSERT_ALL));
      }

      private void setDefaults() {
         int len = players.Count;
         players.Clear();
         fireChanged(new PlayerModelEventArgs(len-1, PlayerModelEventArgs.DELETE_ALL));
      }

   #if WINDOWS_RT || WINDOWS_UWP
      /// <summary>Load STC data from file</summary>
      /// <returns><b>true</b> - successful read; <b>false</b> - not exist or fail read, and set to defaults</returns>
      public async System.Threading.Tasks.Task<bool> Load() {
         var files = await Windows.Storage.ApplicationData.Current.LocalFolder.GetFilesAsync(Windows.Storage.Search.CommonFileQuery.OrderByName);
         var file = files.FirstOrDefault(x => x.Name == StcFile);
         if (file == null) {
            setDefaults();
            return false;
         }

         //file = await Windows.Storage.ApplicationData.Current.LocalFolder.CreateFileAsync(StcFile, Windows.Storage.CreationCollisionOption.OpenIfExists);
         try {
            // 1. read from file
            var content = await Windows.Storage.FileIO.ReadBufferAsync(file);
            byte[] data;
            using (Stream stream = (await file.OpenReadAsync()).AsStreamForRead()) {
               using (BinaryReader input = new BinaryReader(stream, Encoding.UTF8)) {
                  if (input.ReadInt64() != version)
                     throw new Exception("Invalid file data - unknown version");
                  data = new byte[input.ReadInt32()];
                  int read = 0;
                  do {
                     int curr = input.Read(data, read, data.Length - read);
                     if (curr == -1)
                        break;
                     read += curr;
                  } while (read < data.Length);
                  if (read != data.Length)
                     throw new IOException("Invalid data length. Ожидалось " + data.Length + " байт, а прочитано " + read + " байт.");
               }
            }

            // 2. decrypt data
            //data = new Simple3DES(Long.toString(version)).decrypt(data);

            // 3. deserializable object
            using (Stream stream = new MemoryStream()) {
               stream.Write(data, 0, data.Length);
               using (BinaryReader input = new BinaryReader(stream)) {
                  this.readExternal(input);
               }
            }

            return true;
         } catch (Exception ex) {
            System.Diagnostics.Debug.WriteLine(ex.Message);
            setDefaults();
            return false;
         }
      }

      public async void Save() {
         // 1. serializable object
         using (Stream stream = new MemoryStream()) {
            using (BinaryWriter output = new BinaryWriter(stream, Encoding.UTF8)) {
               this.writeExternal(output);
            }

            // 2. crypt data
            byte[] noCryptedData = new byte[stream.Length];
            stream.Position = 0;
            var readed = await stream.ReadAsync(noCryptedData, 0, noCryptedData.Length);
            if (readed != noCryptedData.Length)
               throw new Exception("Not readed saved data");
            //byte[] cryptData = new Simple3DES(Long.toString(version)).encrypt(noCryptedData);
            byte[] cryptData = noCryptedData; // @TODO: temp

            // 3. write to file
            var file = await Windows.Storage.ApplicationData.Current.LocalFolder.CreateFileAsync(StcFile, Windows.Storage.CreationCollisionOption.ReplaceExisting);
            using (var streamFile = await file.OpenStreamForWriteAsync()) {
               using (BinaryWriter output = new BinaryWriter(stream, Encoding.UTF8)) {
                  output.Write(version); // save version and decrypt key
                  int len = cryptData.Length;
                  output.Write(len);
                  output.Write(cryptData);
               }
            }
         }
      }
   #elif WINDOWS_FORMS
      /// <summary>Load STC data from file</summary>
      /// <returns><b>true</b> - successful read; <b>false</b> - not exist or fail read, and set to defaults</returns>
      public bool Load() {
         var file = Assembly.GetExecutingAssembly().GetName().CodeBase;
         if (File.Exists(file)) {
            setDefaults();
            return false;
         }

         try {
            // 1. read from file
            var content = await Windows.Storage.FileIO.ReadBufferAsync(file);
            byte[] data;
            using (Stream stream = (await file.OpenReadAsync()).AsStreamForRead()) {
               using (BinaryReader input = new BinaryReader(stream, Encoding.UTF8)) {
                  if (input.ReadInt64() != version)
                     throw new Exception("Invalid file data - unknown version");
                  data = new byte[input.ReadInt32()];
                  int read = 0;
                  do {
                     int curr = input.Read(data, read, data.Length - read);
                     if (curr == -1)
                        break;
                     read += curr;
                  } while (read < data.Length);
                  if (read != data.Length)
                     throw new IOException("Invalid data length. Ожидалось " + data.Length + " байт, а прочитано " + read + " байт.");
               }
            }

            // 2. decrypt data
            //data = new Simple3DES(Long.toString(version)).decrypt(data);

            // 3. deserializable object
            using (Stream stream = new MemoryStream()) {
               stream.Write(data, 0, data.Length);
               using (BinaryReader input = new BinaryReader(stream)) {
                  this.readExternal(input);
               }
            }

            return true;
         } catch (Exception ex) {
            System.Diagnostics.Debug.WriteLine(ex.Message);
            setDefaults();
            return false;
         }
      }

      public void Save() {
         // 1. serializable object
         using (Stream stream = new MemoryStream()) {
            using (BinaryWriter output = new BinaryWriter(stream, Encoding.UTF8)) {
               this.writeExternal(output);
            }

            // 2. crypt data
            byte[] noCryptedData = new byte[stream.Length];
            stream.Position = 0;
            var readet = await stream.ReadAsync(noCryptedData, 0, noCryptedData.Length);
            if (readet != noCryptedData.Length)
               throw new Exception("Not readet saved data");
            //byte[] cryptData = new Simple3DES(Long.toString(version)).encrypt(noCryptedData);
            byte[] cryptData = noCryptedData;

            // 3. write to file
            var file = await Windows.Storage.ApplicationData.Current.LocalFolder.CreateFileAsync(StcFile, Windows.Storage.CreationCollisionOption.ReplaceExisting);
            using (var streamFile = await file.OpenStreamForWriteAsync()) {
               using (BinaryWriter output = new BinaryWriter(stream, Encoding.UTF8)) {
                  output.Write(version); // save version and decrypt key
                  int len = cryptData.Length;
                  output.Write(len);
                  output.Write(cryptData);
               }
            }
         }
      }
   #else
      public abstract bool Load();
      public abstract void Save();
   #endif

      /// <summary>STatistiCs file name</summary>
      public static string StcFile { get {return "Mines.stc"; } }

      public event PlayerModelChangedHandler OnPlayerChanged = delegate { };
      private void fireChanged(PlayerModelEventArgs e) {
         OnPlayerChanged(this, e);
      }
      public void setUserName(int pos, String name) {
         User user = getUser(pos);
         user.Name = name;
         fireChanged(new PlayerModelEventArgs(pos, PlayerModelEventArgs.UPDATE));
      }

   }

}
